vm/devices/storage/disk_nvme/nvme_driver/fuzz/fuzz_nvme_driver.rs

```diff
@@ -0,0 +1,184 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+//! An interface to fuzz the nvme driver with arbitrary actions
+use crate::fuzz_emulated_device::FuzzEmulatedDevice;
+use crate::get_raw_data;
+
+use arbitrary::{Arbitrary, Unstructured};
+use chipset_device::mmio::ExternallyManagedMmioIntercepts;
+use disk_ramdisk::RamDisk;
+use guestmem::GuestMemory;
+use guid::Guid;
+use nvme::{NvmeController, NvmeControllerCaps};
+use nvme_driver::{Namespace, NvmeDriver};
+use nvme_spec::nvm::DsmRange;
+use pal_async::DefaultDriver;
+use pci_core::msi::MsiInterruptSet;
+use scsi_buffers::OwnedRequestBuffers;
+use std::sync::Arc;
+use user_driver::emulated::DeviceSharedMemory;
+use vmcore::vm_task::{SingleDriverBackend, VmTaskDriverSource};
+
+/// Nvme driver fuzzer
+pub struct FuzzNvmeDriver {
+    driver: Option<NvmeDriver<FuzzEmulatedDevice<NvmeController>>>,
+    namespace: Namespace,
+    payload_mem: GuestMemory,
+}
+
+impl FuzzNvmeDriver {
+    /// Setup a new nvme driver with a fuzz-enabled backend device.
+    pub async fn new(driver: DefaultDriver) -> Self {
+        // Physical storage to back the disk
+        let ram_disk = RamDisk::new(1 << 20, false).unwrap();
+
+        let base_len = 64 << 20;  // 64MB
+        let payload_len = 1 << 20;  // 1MB
+        let mem = DeviceSharedMemory::new(base_len, payload_len);
+
+        // Trasfer buffer
+        let payload_mem = mem
+            .guest_memory()
+            .subrange(base_len as u64, payload_len  as u64, false)
+            .unwrap();
+
+        // Nvme device and driver setup
+        let driver_source =
+            VmTaskDriverSource::new(SingleDriverBackend::new(driver));
+        let mut msi_set = MsiInterruptSet::new();
+        let nvme = NvmeController::new(
+            &driver_source,
+            mem.guest_memory().clone(),
+            &mut msi_set,
+            &mut ExternallyManagedMmioIntercepts,
+            NvmeControllerCaps {
+                msix_count: 2,
+                max_io_queues: 64,
+                subsystem_id: Guid::new_random(),
+            },
+        );
+
+        // Create nvme namespace
+        nvme.client()
+            .add_namespace(1, Arc::new(ram_disk))
```

Comment:
Should this number be randomly generated?

Reply:
See https://github.com/microsoft/openvmm/pull/504#discussion_r1888787992